### ISSUE TITLE
Parsing http-headers in RequestContext - broker query event listener

### ISSUE THREAD

**Issue by tibrewalpratik17** [apache/pinot#12273]
As part of an ongoing discussion in [#12258](https://github.com/apache/pinot/pull/12258#discussion_r1451082198), we would like to gain more inputs from the community on the scope of parsing http-headers in query-request and push it to the event-listener framework. 

For additional context, we at Uber want to track extra metadata info related to query-requests such as caller-name/group, tracing-id, queryhash; alongside other query-level metrics like numDocsScanned and E2E latency (already pushed from our event-listener framework). These metadata info might be easy to identify and aggregate certain query patterns. 
But we want to avoid processing sensitive information like auth-tokens or other authentication-related fields -- this can be a requirement for other companies as well.

Starting this thread to get consensus from the community on the scope of parsing and how best to implement it.

Few solutions which we have thought of:
- Introduce a broker-level config where we specify a **allowlist of headers** to parse from request and push them to request-context object.
- Introducing a broker-level config to specify a **blocklist of headers** to **exclude** from parsing, pushing all others to the request-context object.

cc @walterddr @ankitsultana

**Comment by walterddr** [comment]
i think an allowed list would be desirable. but it is a bit hard to make it configurable/extensible without overriding it. do we have any idea on how one can extend this to support extra properties?

**Comment by walterddr** [comment]
some example i had in mind, 
- should we allow cluster config modification (e.g. not as a static broker conf file but modifiable during runtime)
- where do we allow this config key? event listener scope (of the event listener factory) and tracer scope (request context tracer) both might be a good scope to put as prefix;

**Comment by tibrewalpratik17** [comment]
> i think an allowed list would be desirable. but it is a bit hard to make it configurable/extensible without overriding it. do we have any idea on how one can extend this to support extra properties? should we allow cluster config modification (e.g. not as a static broker conf file but modifiable during runtime)

We need this info at `RequestHandler` class when pushing tracked-headers to `RequestContext`. 
To make it configurable:
- we will either have to recreate `RequestHandler` classes if this list is passed as a param. 
- store it as a static reference somewhere and create a controller API to update these properties for a set of component. The default value can be read from that component's config file. 
But after updating, we might need to push the updated list to some conf file again so that restarts don't affect the list?

**Comment by walterddr** [comment]
so there's 3 
- requires no restart --> pick up immediately on next query
- requires restart 
- require changing the image (or modify the files shipped together with the deployment packaging)

ideally if we don't need to go through step 3 that would be perfect; i dont think we need to go with no restart though

**Comment by tibrewalpratik17** [comment]
for step 2 `requires restart`, we might have to store this info in ZK then, is that okay?

**Comment by walterddr** [comment]
that's fine. looks like cluster config is saved in zk

**Comment by walterddr** [comment]
> This will be a Map<String, List>. The map can have pinot-components as keys - pinot.server, pinot.broker, pinot.controller

why would we do differently for different components? 

> During cluster starter / component restarts, we will fetch values from both conf and ZK; and do a union-distinct of both the lists.

let's make sure which overrides what, my argument is cluster level (in zk) should be override by component conf file
e.g. 
1. first load the headers listed in the cluster zk conf;
2. then load the component level conf 
3. union both; (could you elaborate on the union-distinct? i can only think of adding instead of removing)

**Comment by tibrewalpratik17** [comment]
> why would we do differently for different components?

Because pinot-server / broker can have different headers for the same query but yeah we can have one general list for all components as well.

> union both; (could you elaborate on the union-distinct? i can only think of adding instead of removing)

This i proposed just in case we want to remove some header due to any manual misconfiguration or anything. In the case, if ZK has it and conf-file does not; doing a union will pick the header again. Better to expose an API for removing from ZK then.

So to summarise the entire flow:

- first load the headers listed in the cluster zk conf
- then load the component level conf
- do a union

To add a new header, use an controller API to add it to ZK conf. Restart and then it will be picked up -- no need to necessarily change the conf file here.
To remove a header (just in case), use controller API to remove from ZK conf. Remove from conf file as well. Restart.
